useContext.md finished (#560)
reactjs/uk.react.dev: src/content/reference/react/useContext.md

<Intro>

> `useContext` це хук, який дозволяє читати [контекст](/learn/passing-data-deeply-with-context) компонента та підписуватись на нього.

Reviewer: ```suggestion
`useContext` це хук, який дозволяє читати [контекст](/learn/passing-data-deeply-with-context) компонента та підписуватися на нього.
```

  // ...
```

[Дивіться інші приклади.](#usage)

#### Параметри {/*parameters*/}

* `SomeContext`: це об'єкт контексту, який було створено за допомогою функції [`createContext`](/reference/react/createContext). Контекст безпосередньо не містить інформацію, він лише визначає що ви можете передати або отримати з компонентів.

#### Повернення {/*returns*/}

> Викликаючи `useContext`, ви отримуєте значення (`value`), яке відповідає контексту компонента, що викликає цей хук. Це значення визначається як `value`, передане найближчому `SomeContext.Provider`, розташованому вище в ієрархії компонентів від компонента, що використовує цей контекст. Якщо такого провайдера немає, повертається значення `defaultValue`, яке ви передали функції [`createContext`](/reference/react/createContext). Повернуте значення завжди актуальне. React автоматично повторно рендерить усі компоненти, що використовують контекст, якщо значення контексту змінюється.

Reviewer: ```suggestion
`useContext` повертає значення контексту для компонента, що викликає цей хук. Це значення визначається як `value`, передане найближчому `SomeContext.Provider`, розташованому вище деревом відносно поточного компонента. Якщо такого провайдера немає, повертається значення `defaultValue`, яке ви передали функції [`createContext`](/reference/react/createContext). Повернуте значення завжди актуальне. React автоматично повторно рендерить усі компоненти, що використовують контекст, якщо значення контексту змінюється.
```
Краще простіше. У документації і без того дуже багато розжовується

#### Особливості {/*caveats*/}

> * Виклик `useContext()` не бачить значення контексту, яке визначається або змінюється у *цьому ж* компоненті. Відповідний `<Context.Provider>` **повинен бути розташований *вище*** компонента, що викликає `useContext()`.

Reviewer: ```suggestion
* Eлементи-провайдери не впливають на виклик `useContext()` у компоненті, з якого й повертаються. Відповідний `<Context.Provider>` **повинен бути розташований *вище*** компонента, що викликає `useContext()`.
```

* React **автоматично оновлює** всі дочірні компоненти, які використовують певний контекст, починаючи з провайдера, що отримує змінене значення `value`. Попереднє та наступне значення порівнюються за допомогою [`Object.is`](https://webdoky.org/uk/docs/Web/JavaScript/Reference/Global_Objects/Object/is/). Пропуск повторних рендерів за допомогою [`memo`](/reference/react/memo) не заважає дочірнім компонентам отримувати оновлене значення контексту.

> * Якщо ваша система збірки створює дублікати модулів у вихідному коді (що може статися через симлінки), це може порушити контекст. Передача через контекст працює тільки у випадку, якщо `SomeContext`, що використовується для надання контексту, та `SomeContext`, для його зчитування, є ***точно* тим самим об'єктом**, що визначається порівнянням `===`.

Reviewer: ```suggestion
* Якщо ваша система збірки створює дублікати модулів у вихідному коді (що може статися через символьні посилання), це може порушити контекст. Передавання через контекст працює тільки у разі, якщо `SomeContext` для надання контексту та `SomeContext` для його зчитування є ***точно* тим самим об'єктом**, що визначається порівнянням `===`.
```
- [символьні посилання](https://uk.wikipedia.org/wiki/%D0%A1%D0%B8%D0%BC%D0%B2%D0%BE%D0%BB%D1%8C%D0%BD%D0%B5_%D0%BF%D0%BE%D1%81%D0%B8%D0%BB%D0%B0%D0%BD%D0%BD%D1%8F)
- [передавання](https://onlinecorrector.com.ua/uk/%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D0%B2%D0%B0%D0%BD%D0%BD%D1%8F/)
- [у разі](https://onlinecorrector.com.ua/uk/%D1%83-%D1%80%D0%B0%D0%B7%D1%96/)

}
```

Як наслідок цієї зміни, навіть якщо `MyApp` потребує повторного рендеру, компоненти, що викликають `useContext(AuthContext)`, не потребуватимуть, якщо `currentUser` не змінився.

Дізнайтеся більше про [`useMemo`](/reference/react/useMemo#skipping-re-rendering-of-components) та [`useCallback`](/reference/react/useCallback#skipping-re-rendering-of-components).

---

> ## Вирішення проблем {/*troubleshooting*/}

Reviewer: ```suggestion
## Усунення несправностей {/*troubleshooting*/}
```

### Компонент не бачить значення з провайдера {/*my-component-doesnt-see-the-value-from-my-provider*/}

Існує кілька поширених причин, чому це може статися:

1. Ви розміщуєте `<SomeContext.Provider>` у тому ж компоненті або нижче, ніж компонент, де викликано `useContext()`. Перемістіть `<SomeContext.Provider>` *вище і зовні* компонента, який викликає `useContext()`.

> 2. Можливо, ви забули обгорнути свій компонент у `<SomeContext.Provider>`, або розмістили його в іншій частині дерева, ніж планували. Переконайтесь за допомогою [React DevTools](/learn/react-developer-tools), що ієрархія компонентів налаштована правильно.

Reviewer: ```suggestion
2. Можливо, ви забули обгорнути свій компонент у `<SomeContext.Provider>` або розмістили його в іншій, ніж задумали, частині дерева. Переконайтеся за допомогою [React DevTools](/learn/react-developer-tools), що ієрархія компонентів налаштована правильно.
```

> 3. Можливо, ви зіткнулися з проблемою збірки, через яку `SomeContext`, доступний у провайдері, і `SomeContext`, що використовується споживачем, є різними об'єктами. Це може статися, наприклад, якщо ви використовуєте псевдоніми. Можете перевірити це, присвоївши їх глобальним змінним, як-от `window.SomeContext1` і `window.SomeContext2`, а потім порівняти `window.SomeContext1 === window.SomeContext2` в консолі. Якщо вони не тотожні, виправте цю проблему на рівні збірки.

Reviewer: ```suggestion
3. Можливо, ви зіткнулися з проблемою збірки, через яку `SomeContext` із компонента-провайдера і `SomeContext` компонента-читача є різними об'єктами. Це може статися, наприклад, якщо ви використовуєте символьні посилання. Можете перевірити це, присвоївши їх глобальним змінним, як-от `window.SomeContext1` і `window.SomeContext2`, а потім порівняти `window.SomeContext1 === window.SomeContext2` в консолі. Якщо вони не тотожні, виправте цю проблему на рівні збірки.
```

> ### Я завжди отримую `undefined` з контексту, хоча значення за замовчуванням інше {/*i-am-always-getting-undefined-from-my-context-although-the-default-value-is-different*/}

Reviewer: ```suggestion
### Я завжди отримую `undefined` із контексту, хоча початкове значення інше {/*i-am-always-getting-undefined-from-my-context-although-the-default-value-is-different*/}
```

Можливо, у вас є провайдер без `value` у дереві:

```js {1,2}

> // 🚩 Не працює: немає властивості value

Reviewer: ```suggestion
// 🚩 Не працює: немає пропа value
```
для властивостей react-компонентів використовуємо саме проп

<ThemeContext.Provider>
   <Button />
</ThemeContext.Provider>
```

> Якщо ви забули вказати `value`, це еквівалентно передачі `value={undefined}`.

Reviewer: ```suggestion
Якщо ви забули вказати `value`, це еквівалентно передаванню `value={undefined}`.
```

> Також, можливо, ви випадково застосували помилкову назву властивості:

Reviewer: ```suggestion
Також, можливо, ви випадково застосували помилкову назву пропа:
```

```js {1,2}

> // 🚩 Не працює: властивість має називатися "value"

Reviewer: ```suggestion
// 🚩 Не працює: проп має називатися "value"
```

<ThemeContext.Provider theme={theme}>
   <Button />
</ThemeContext.Provider>
```

> В обох випадках React покаже попередження у консолі. Щоб виправити це, назвіть властивість `value`:

Reviewer: ```suggestion
В обох випадках React покаже попередження у консолі. Щоб виправити це, назвіть проп `value`:
```

```js {1,2}

> // ✅ Передача властивості value

Reviewer: ```suggestion
// ✅ Передача пропа value
```

<ThemeContext.Provider value={theme}>
   <Button />
</ThemeContext.Provider>
```

> Зверніть увагу, що [значення за замовчуванням, яке ви передали у `createContext(defaultValue)`](#specifying-a-fallback-default-value) застосовується **тільки якщо вище в ієрархії не знайдено жодного провайдера.** Якщо в будь-якому місці дерева батьківських компонентів є `<SomeContext.Provider value={undefined}>`, компонент, що викликає `useContext(SomeContext)`, *отримає* `undefined` як значення контексту.

Reviewer: ```suggestion
Зверніть увагу, що [початкове значення, яке ви передали у `createContext(defaultValue)`](#specifying-a-fallback-default-value), застосовується, **тільки якщо вище в ієрархії не знайдено жодного провайдера.** Якщо в будь-якому місці дерева батьківських компонентів є `<SomeContext.Provider value={undefined}>`, компонент, що викликає `useContext(SomeContext)`, *отримає* `undefined` як значення контексту.
```
